python/pyspark/ml/torch/deepspeed/deepspeed_distributor.py

```diff
@@ -0,0 +1,144 @@
+
+            nnodes: int
+                The number of nodes that should be used for the run.
+
+            local_mode: bool
+                Whether or not to run the training in a distributed fashion or just locally.
+
+            use_gpu: bool
+                Boolean flag to determine whether to utilize gpus.
+
+            deepspeed_config: Union[Dict[str,Any], str] or None:
+                The configuration file to be used for launching the deepspeed application. 
+                If it is a dictionary mapping parameters to values, then we will create the file.
+                If None, deepspeed will fall back to default parameters.
+        """
+        num_processes = num_gpus * nnodes
+        super().__init__(num_processes, local_mode, use_gpu)
+        self.deepspeed_config = deepspeed_config 
+        self.ssl_conf = "deepspeed.spark.distributor.ignoreSsl"
+        self._validate_input_params()
+        self.input_params = self._create_input_params()
+        self.cleanup_deepspeed_conf = False
+
+    @staticmethod
+    def _get_deepspeed_config_path(deepspeed_config):
+        if isinstance(deepspeed_config, dict):
+            with tempfile.NamedTemporaryFile(mode='w+', delete=False, suffix='.json') as fil:
+                json.dump(deepspeed_config, fil)
+                return fil.name
+        deepspeed_config_path = deepspeed_config
+        # Empty value means the deepspeed will fall back to default settings.
+        if deepspeed_config == None:
+            return "" 
+        return deepspeed_config_path
+
+
+    @staticmethod 
+    def _get_torchrun_args(local_mode: bool, num_processes: int) -> Tuple[List[Any], int]:
+        """
+        Given the mode and the number of processes, create the arguments to be given to deepspeed
+        
+        Parameters
+        ---------
+        local_mode: bool
+            Whether or not we are running training locally or in a distributed fashion
+
+        num_processes: int
+            The number of processes that we are going to use (number of gpus per node * number of nodes)
+
+        Returns
+        ------
+        Tuple[List[Any], int]
+            A tuple containing a list of arguments to pass as pytorch args to deepspeed, as well as the number of processes per node
+        """
+        if local_mode:
+            torchrun_args = ["--standalone","--nnodes=1"]
+            processes_per_node = num_processes
+            return torchrun_args, processes_per_node
+
+        master_addr = os.environ["MASTER_ADDR"]
+        master_port = os.environ["MASTER_PORT"]
+        node_rank = os.environ["RANK"]
+        torchrun_args = [
+            f"--nnodes={num_processes}",
+            f"--node_rank={node_rank}",
+            f"--rdzv_endpoint={master_addr}:{master_port}",
+            "--rdzv_id=0",
+        ]
+        processes_per_node = 1
+        return torchrun_args, processes_per_node
```

Comment:
It looks like these are the same args created in TorchDistributor's _create_torchrun_command. Let's move this function to that class and call it from _create_torchrun_command

Reply:
Good call - moved it now!